# Multimoal Parsing with LlamaParse

## 0. Import Library & Set Environment Variables

In [15]:
from dotenv import load_dotenv
import nest_asyncio

from glob import glob
import os

from llama_parse import LlamaParse

from autorag.parser import Parser

import pandas as pd

In [5]:
load_dotenv()

True

In [6]:
nest_asyncio.apply()

In [21]:
cur_dir = os.getcwd()
root_dir = cur_dir
data_dir = os.path.join(root_dir, "data", "parse_multimodal")
glob_path = os.path.join(data_dir, "*")

file_path = glob(glob_path)
file_path

['/home/martinus/workspace/RAG-Evaluation/Parsing/data/parse_multimodal/Llama_image_text.pdf',
 '/home/martinus/workspace/RAG-Evaluation/Parsing/data/parse_multimodal/Llama_image.pdf']

## 1. Parsing with Llama Parse

model list
- openai-gpt4o
- openai-gpt-4o-mini
- anthropic-sonnet-3.5
- gemini-1.5-flash
- gemini-1.5-pro

In [ ]:
parse_instance = LlamaParse(
    results_type="markdown",
    language="ko",
    use_vendor_multimodal_model=True,
    vendor_multimodal_model="openai-gpt-4o-mini"
    #vendor_multimodal_api_key="..."
)

In [23]:
parse_instance.load_data(file_path)

Parsing files:   0%|          | 0/2 [00:00<?, ?it/s]

[04/27/25 21:26:05] INFO     [_client.py:1786] >> HTTP Request:  ]8;id=495931;file:///home/martinus/llm/lib/python3.10/site-packages/httpx/_client.py\_client.py]8;;\:]8;id=373497;file:///home/martinus/llm/lib/python3.10/site-packages/httpx/_client.py#1786\1786]8;;\
                             POST                                               
                             https://api.cloud.llamaindex.ai/api                
                             /parsing/upload "HTTP/1.1 200 OK"                  
                    INFO     [_client.py:1786] >> HTTP Request:  ]8;id=699988;file:///home/martinus/llm/lib/python3.10/site-packages/httpx/_client.py\_client.py]8;;\:]8;id=681326;file:///home/martinus/llm/lib/python3.10/site-packages/httpx/_client.py#1786\1786]8;;\
                             POST                                               
                             https://api.cloud.llamaindex.ai/api                
                             /parsing/upload "H

Parsing files:  50%|█████     | 1/2 [00:06<00:06,  6.84s/it]

[04/27/25 21:26:12] INFO     [_client.py:1786] >> HTTP Request:  ]8;id=383765;file:///home/martinus/llm/lib/python3.10/site-packages/httpx/_client.py\_client.py]8;;\:]8;id=599670;file:///home/martinus/llm/lib/python3.10/site-packages/httpx/_client.py#1786\1786]8;;\
                             GET                                                
                             https://api.cloud.llamaindex.ai/api                
                             /parsing/job/6be58682-92bd-4c95-908                
                             e-1b044d8f54c2 "HTTP/1.1 200 OK"                   
[04/27/25 21:26:16] INFO     [_client.py:1786] >> HTTP Request:  ]8;id=30441;file:///home/martinus/llm/lib/python3.10/site-packages/httpx/_client.py\_client.py]8;;\:]8;id=180310;file:///home/martinus/llm/lib/python3.10/site-packages/httpx/_client.py#1786\1786]8;;\
                             GET                                                
                             https://api.cloud.l

Parsing files: 100%|██████████| 2/2 [00:38<00:00, 19.36s/it]


[Document(id_='486cffa4-0fe8-4ae2-98a6-b32d6ae31c25', embedding=None, metadata={}, excluded_embed_metadata_keys=[], excluded_llm_metadata_keys=[], relationships={}, text='# Introducing Llama 3.1: Our most capable models to date\n\nAs our largest model yet, training Llama on 405B on over 16 trillion tokens was a major challenge. To enable training runs at this scale and achieve the results we have in a reasonable amount of time, we significantly optimized our full training stack and pushed our model training to over 16 thousand H100 GPUs, making the 405B the first Llama model trained at this scale.\n\n```\nINPUT\nText tokens\n```\n→\n```\nToken embeddings\n```\n→\n```\nSelf-attention\n```\n→\n```\nFeedforward network\n```\n→\n...\n→\n```\nSelf-attention\n```\n→\n```\nFeedforward network\n```\n→\n```\nOUTPUT\nText token\n```\n\n*AUTOREGRESSIVE DECODING*\n\nTo address this, we made design choices that focus on keeping the model development process scalable and straightforward.\n\n- We opt

## 2. Parsing with AutoRAG

In [29]:
pjt_dir = os.path.join(root_dir, "multimodal")
parser = Parser(data_path_glob=glob_path, project_dir=pjt_dir)

In [30]:
yaml_path = os.path.join(root_dir, "config", "multimodal.yaml")
parser.start_parsing(yaml_path, all_files=True)

[04/27/25 21:28:54] INFO     [parser.py:29] >> Parsing Start...     ]8;id=650032;file:///home/martinus/llm/lib/python3.10/site-packages/autorag/parser.py\parser.py]8;;\:]8;id=46662;file:///home/martinus/llm/lib/python3.10/site-packages/autorag/parser.py#29\29]8;;\
                    INFO     [base.py:23] >> Running parser -         ]8;id=520455;file:///home/martinus/llm/lib/python3.10/site-packages/autorag/data/parse/base.py\base.py]8;;\:]8;id=464332;file:///home/martinus/llm/lib/python3.10/site-packages/autorag/data/parse/base.py#23\23]8;;\
                             llama_parse module...                              


[04/27/25 21:28:57] INFO     [_client.py:1786] >> HTTP Request:  ]8;id=572049;file:///home/martinus/llm/lib/python3.10/site-packages/httpx/_client.py\_client.py]8;;\:]8;id=680111;file:///home/martinus/llm/lib/python3.10/site-packages/httpx/_client.py#1786\1786]8;;\
                             POST                                               
                             https://api.cloud.llamaindex.ai/api                
                             /parsing/upload "HTTP/1.1 200 OK"                  
Started parsing the file under job_id f40223ff-3d46-44e2-a4cb-0f693b0e1118
[04/27/25 21:28:58] INFO     [_client.py:1786] >> HTTP Request:  ]8;id=309595;file:///home/martinus/llm/lib/python3.10/site-packages/httpx/_client.py\_client.py]8;;\:]8;id=455829;file:///home/martinus/llm/lib/python3.10/site-packages/httpx/_client.py#1786\1786]8;;\
                             POST                                               
                             https://api.cloud.llamai

## 3. Check Result

In [31]:
result_path = os.path.join(pjt_dir, 'parsed_result.parquet')
result = pd.read_parquet(result_path)
result

texts  \
0  # Introducing Llama 3.1: Our most capable mode...   
1  | Category     | Benchmark                  | ...   

                                                path  page  \
0  /home/martinus/workspace/RAG-Evaluation/Parsin...     1   
1  /home/martinus/workspace/RAG-Evaluation/Parsin...     1   

  last_modified_datetime  
0             2025-04-27  
1             2025-04-27

In [32]:
print(result['texts'].tolist()[0])

# Introducing Llama 3.1: Our most capable models to date

As our largest model yet, training Llama on 405B on over 16 trillion tokens was a major challenge. To enable training runs at this scale and achieve the results we have in a reasonable amount of time, we significantly optimized our full training stack and pushed our model training to over 16 thousand H100 GPUs, making the 405B the first Llama model trained at this scale.

## Model Architecture

### Training Process

- **Input**: Text tokens
- **Token embeddings**
- **Self-attention**
- **Feedforward network**
- **Self-attention**
- **Feedforward network**
- **Output**: Text token
- **Autoregressive Decoding**

To address this, we made design choices that focus on keeping the model development process scalable and straightforward.

- We opted for a standard decoder-only transformer model architecture with minor adaptations rather than a mixture-of-experts model to maximize training stability.
- We adopted an iterative post-traini

In [33]:
print(result['texts'].tolist()[1])

| Category     | Benchmark                  | Llama 3.1 88B      | Gemma 2 9B IT | Mistral 7B Instruct | Llama 3.1 70B | Mixtral 8x228 Instruct | GPT 3.5 Turbo |      |
| ------------ | -------------------------- | ------------------ | ------------- | ------------------- | ------------- | ---------------------- | ------------- | ---- |
| General      | MMLU (0-shot, CoT)         | 73.0               | 72.3          | 60.5                | 86.0          | 79.9                   | 69.8          |      |
|              | MMLU PRO (5-shot, non-CoT) | 48.3               | 36.9          | 36.9                | 66.4          | 56.3                   | 49.2          |      |
|              | IFEval                     | 80.4               | 73.6          | 57.6                | 87.5          | 72.7                   | 69.9          |      |
|              | Code                       | HumanEval (0-shot) | 72.6          | 54.3                | 40.2          | 80.5                   | 78.6     

| Category     | Benchmark                  | Llama 3.1 88B      | Gemma 2 9B IT | Mistral 7B Instruct | Llama 3.1 70B | Mixtral 8x228 Instruct | GPT 3.5 Turbo |      |
| ------------ | -------------------------- | ------------------ | ------------- | ------------------- | ------------- | ---------------------- | ------------- | ---- |
| General      | MMLU (0-shot, CoT)         | 73.0               | 72.3          | 60.5                | 86.0          | 79.9                   | 69.8          |      |
|              | MMLU PRO (5-shot, non-CoT) | 48.3               | 36.9          | 36.9                | 66.4          | 56.3                   | 49.2          |      |
|              | IFEval                     | 80.4               | 73.6          | 57.6                | 87.5          | 72.7                   | 69.9          |      |
|              | Code                       | HumanEval (0-shot) | 72.6          | 54.3                | 40.2          | 80.5                   | 78.6          | 82.0 |
| Math         | MBPP EvalPlus (0-shot)     | 72.8               | 71.7          | 49.5                | 86.0          | 78.6                   | 82.0          |      |
|              | GSM8K (0-shot, CoT)        | 84.5               | 76.7          | 53.2                | 95.1          | 88.2                   | 81.6          |      |
|              | MATH (0-shot, CoT)         | 51.9               | 44.3          | 13.0                | 68.0          | 54.1                   | 43.1          |      |
| Reasoning    | ARC Challenge (0-shot)     | 83.4               | 87.6          | 74.2                | 94.8          | 88.7                   | 83.7          |      |
|              | GPAO (0-shot, CoT)         | 32.8               | 28.8          | 46.7                | 33.3          | 30.8                   | -             |      |
| Tool use     | BFCL                       | 76.1               | 60.4          | 84.8                | -             | 85.9                   | -             |      |
| Long context | ZeroSCROLLS/QUALITY        | 81.0               | -             | -                   | 90.5          | -                      | -             |      |
| -            | InfiniteBench/Inf/MC       | 98.8               | -             | -                   | 78.2          | -                      | -             |      |
| -            | NIN/Multi-needle           | -                  | -             | -                   | 97.5          | -                      | -             |      |
| Multilingual | Multilingual MGSM (0-shot) | 68.9               | 53.2          | 29.9                | 86.9          | 71.1                   | 51.4          |      |

| Comparison                           | Win   | Tie   | Loss  |
| ------------------------------------ | ----- | ----- | ----- |
| Llama 3.1 405B vs GPT-4-0125-Preview | 23.3% | 52.2% | 24.5% |
| Llama 3.1 405B vs GPT-4o             | 19.1% | 51.7% | 29.2% |
| Llama 3.1 405B vs Claude 3.5 Sonnet  | 24.9% | 50.8% | 24.2% |